Sending Error Emails

jacob6838 opened the pull request:
- Creating error log handler to send emails to subscribed recipients (see screenshot below)
    - Emails feature error message, stack trace, timestamp, link to cloud run logs, and email link to admin user to unsubscribe
- Admin user page allows admins to set which users receive emails (see screenshot below)


User Checkbox: 
![image](https://github.com/CDOT-CV/jpo-cvmanager/assets/51723791/11700649-76ed-4902-8332-420376432cae)

Email Example: 
![image](https://github.com/CDOT-CV/jpo-cvmanager/assets/51723791/b4c2e6c4-8bc5-4bcf-b63d-0958dde09279)

jacob6838 committed:
Creating error_email log handler and adding UI components

jacob6838 committed:
Updating error email environment variables

drewjj commented on api/sample.env:
```diff
@@ -14,4 +14,8 @@ CSM_EMAIL_APP_USERNAME=
+CSM_TARGET_SMTP_SERVER_PORT=
+
+# Error Email Contact Configuration
+CLOUD_RUN_LOGS_LINK=https://console.cloud.google.com/run/detail/us-central1/rsu-manager-cloud-run-api/logs?authuser=1&project=cdot-oim-cv-dev
```
Lets name this CLOUD_LOGS_LINK or something. We don't plan to run this in GCP Cloud Run for much longer and have been trying to get rid of dependencies to GCP. This isn't really a dependency to Cloud Run to my understanding so it would be good to not make it sound like that.

jacob6838 replied:
Updated to LOGS_LINK

drewjj commented on api/src/error_email/error_email_template.html:
```diff
@@ -0,0 +1,62 @@
+    <br />
+    <p>Error occurred at: ##_ERROR_TIME_##</p>
+    <br />
+    <p>View this error in Cloud Run logs: <a href="##_CLOUD_RUN_LOGS_LINK_##">rsu-manager-cloud-run-api logs</a></p>
```
Lets make this less dependent on Cloud Run. I think any link to any logs would work since it is just a URL.

jacob6838 replied:
Updated to LOGS_LINK

drewjj reviewed (changes requested):
Looks great. I still need to test this for myself but these are my main comments at the moment based off just the code additions.

payneBrandon commented on api/src/smtp_error_handler.py:
```diff
@@ -0,0 +1,82 @@
+                
+            body_content = open("./error_email/error_email_template.html").read()
+            EMAIL_KEYS = {
+                'ENVIRONMENT': get_environment_name(os.getenv('INSTANCE_CONNECTION_NAME', 'cdot-oim-cv-dev:us-west3:rsu-manager')),
```
Rather than having the system default to our development environment if not found, maybe we should let the exception pop through. These values (minus the email since that can have a sensible default) seem to be required, so maybe we do a check at the beginning of the class and throw if not found? That way we'd see the error when deploying and not accidentally send something to dev.

jacob6838 replied:
Updated to using os.environ instead getenv

payneBrandon commented on api/src/error_email/error_email_template.html:
```diff
@@ -0,0 +1,62 @@
+    <br />
+    <p>--</p>
+    <p>
+      If you no longer wish to receive these emails, please uncheck the "receive error emails" checkbox on the CV
```
Doesn't need to be in this PR, but I think future-state here will be an "unsubscribe" link that directly takes the user out of the email list.

jacob6838 replied:
Yeah, agree. Maybe we could create an unauthenticated endpoint to the API that allows user to unsubscribe?

jacob6838 replied:
Added an unauthenticated unsubscribe endpoint

payneBrandon reviewed (changes requested):
Running into trouble with the addons unit tests running.

jacob6838 committed:
Adding unsubscribe endpoint, updating environment variables

jacob6838 commented:
I just added an unauthenticated unsubscribe endpoint, and added a link to unsubscribe a user in each email. This also means that now, we send 1 email per user, because they each have a unique unsubscribe endpoint. The url looks like: https://unsubscribe---rsu-manager-cloud-run-api-kcrcl2kkaq-uc.a.run.app/unsubscribe-error-emails/jfrye@neaeraconsulting.com

payneBrandon reviewed (changes requested):
Looking good! Let's get some unit tests in to cover the new functionality in smtp_error_handler and unsub_error_emails and we'll be good to go!

jacob6838 committed:
Adding unit tests and fixing send emails

jacob6838 commented:
95% code coverage for smtp_error_handler and 100% on unsub_error_emails

jacob6838 committed:
Merge branch 'develop' into error-emails

drewjj reviewed (approved):
Looks good to me!

payneBrandon commented on api/src/unsub_error_emails.py:
```diff
@@ -0,0 +1,40 @@
+            resp_code = unsubscribe_user(email)
+
+            if resp_code == 404:
+                return (f"User with email {email} does not exist", 404, self.headers)
```
Rather than a 404, which would indicate the resource wasn't found on our server, should we return something else? Maybe a 400 bad request?

jacob6838 replied:
Just changed this code to 400

payneBrandon reviewed (changes requested):
Running unit tests currently fails: 

![image](https://github.com/CDOT-CV/jpo-cvmanager/assets/2932159/e379635a-703c-44ac-ad73-0ba2bfdf8ec1)

drewjj reviewed (changes requested):
Alright, I admit it. I didn't run the tests lol. I am revoking my approval because I can confirm the tests for the API do not pass. Looks like the webapp tests also fail on 5 tests.

jacob6838 committed:
Updating snapshot tests

jacob6838 committed:
switching 404 to 400

jacob6838 committed:
switching 404 to 400 pt 2

jacob6838 committed:
Updating unit tests to use 400 resp code

jacob6838 commented:
The API tests are failing on develop. If you set the CORS_DOMAIN environment variable, then they all succeed
```
$env:CORS_DOMAIN="test.com"
```

jacob6838 committed:
Merge branch 'develop' into error-emails

jacob6838 committed:
Updating unit tests

jacob6838 commented:
All merge conflicts have been resolved, and all unit tests pass!

payneBrandon commented on services/api/sample.env:
```diff
@@ -14,4 +14,8 @@ CSM_EMAIL_APP_USERNAME=
+CSM_TARGET_SMTP_SERVER_PORT=
+
+# Error Email Contact Configuration
+CLOUD_RUN_LOGS_LINK=https://console.cloud.google.com/run/detail/us-central1/rsu-manager-cloud-run-api/logs?authuser=1&project=cdot-oim-cv-dev
```
Since this is the sample.env should we remove these values?

jacob6838 replied:
I go both ways. Sometimes it is nice if the sample.env has samples of non-secret values. But that can be confusing to some people, because they will have to change it still. I will remove it

jacob6838 replied:
Just updated the rest of the sample.env as well, it was a little out of date

jacob6838 committed:
Updating sample.env and environment_name env variable

jacob6838 committed:
Fixing unit tests

jacob6838 committed:
Update smtp_error_handler.py

drewjj reviewed (commented):
Sounds like you may be working on it but the Python unit tests are not passing.

jacob6838 commented:
> Sounds like you may be working on it but the Python unit tests are not passing.

Wait, they aren't? I just ran: 
```
python -m pytest
```
from services/api and git all greens

drewjj commented:
> > Sounds like you may be working on it but the Python unit tests are not passing.
> 
> Wait, they aren't? I just ran:
> 
> ```
> python -m pytest
> ```
> 
> from services/api and git all greens

They actually need to run from services, not api. Just to conform with the vscode tasks and our other tests. It seems to be imports failing is all.

jacob6838 committed:
Updating unit tests to run from services dir

jacob6838 commented:
> > > Sounds like you may be working on it but the Python unit tests are not passing.
> > 
> > 
> > Wait, they aren't? I just ran:
> > ```
> > python -m pytest
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > from services/api and git all greens
> 
> They actually need to run from services, not api. Just to conform with the vscode tasks and our other tests. It seems to be imports failing is all.

Just updated the unit tests, they succeed now!

drewjj reviewed (approved):
Looks great to me now!

jacob6838 committed:
Merge branch 'develop' into error-emails

payneBrandon reviewed (approved):
lgtm!
